File: src/data-feed.js
const config = require('./config');

let _yf = null;
async function yf() {
  if (!_yf) {
    const mod = await import('yahoo-finance2');
    const YahooFinance = mod.default;
    _yf = new YahooFinance();
  }
  return _yf;
}

class DataFeed {
  constructor() {
    this._dailyCache = {};
    this._dailyTs = 0;
    this._weeklyCache = {};
    this._weeklyTs = 0;
  }

  async fetchIntraday() {
    const out = {};
    const period1 = new Date();
    period1.setDate(period1.getDate() - 5);

    const yahooFinance = await yf();
    for (const ticker of config.TICKERS) {
      try {
        const res = await yahooFinance.chart(ticker, {
          period1,
          interval: '1m',
          includePrePost: true,
        });
        if (res?.quotes?.length) out[ticker] = res.quotes;
      } catch (e) {
        console.warn(`Intraday ${ticker}:`, e.message);
      }
    }
    return out;
  }

  async fetchDaily() {
    if (Date.now() - this._dailyTs < 300_000) return this._dailyCache;
    const out = {};
    const period1 = new Date();
    period1.setDate(period1.getDate() - 30);

    const yahooFinance = await yf();
    for (const ticker of config.TICKERS) {
      try {
        const res = await yahooFinance.chart(ticker, { period1, interval: '1d' });
        if (res?.quotes?.length) out[ticker] = res.quotes;
      } catch (e) {
        console.warn(`Daily ${ticker}:`, e.message);
      }
    }
    this._dailyCache = out;
    this._dailyTs = Date.now();
    return out;
  }

  async fetchWeekly() {
    if (Date.now() - this._weeklyTs < 1_800_000) return this._weeklyCache;
    const out = {};
    const period1 = new Date();
    period1.setDate(period1.getDate() - 90);

    const yahooFinance = await yf();
    for (const ticker of config.TICKERS) {
      try {
        const res = await yahooFinance.chart(ticker, { period1, interval: '1wk' });
        if (res?.quotes?.length) out[ticker] = res.quotes;
      } catch (e) {
        console.warn(`Weekly ${ticker}:`, e.message);
      }
    }
    this._weeklyCache = out;
    this._weeklyTs = Date.now();
    return out;
  }

  async fetchAll() {
    const [intraday, daily, weekly] = await Promise.all([
      this.fetchIntraday(),
      this.fetchDaily(),
      this.fetchWeekly(),
    ]);
    return { intraday, daily, weekly };
  }
}

module.exports = DataFeed;

File: data_feed.py
"""Yahoo Finance data feed with caching."""

import logging
from datetime import datetime

import pandas as pd
import pytz
import yfinance as yf

from config import TICKERS, TIMEZONE

logger = logging.getLogger(__name__)
ET = pytz.timezone(TIMEZONE)


class DataFeed:
    def __init__(self):
        self._daily_cache: dict[str, pd.DataFrame] = {}
        self._weekly_cache: dict[str, pd.DataFrame] = {}
        self._daily_ts: datetime | None = None
        self._weekly_ts: datetime | None = None

    # ------------------------------------------------------------------
    def fetch_intraday(self) -> dict[str, pd.DataFrame]:
        """1-min candles for the last 5 days (max 7d for 1m on yfinance)."""
        out: dict[str, pd.DataFrame] = {}
        for ticker in TICKERS:
            try:
                df = yf.Ticker(ticker).history(
                    period="5d", interval="1m", prepost=True
                )
                if df.empty:
                    continue
                # Ensure ET timezone
                if df.index.tz is not None:
                    df.index = df.index.tz_convert(ET)
                else:
                    df.index = df.index.tz_localize(ET)
                out[ticker] = df
            except Exception as e:
                logger.warning("Intraday fetch failed for %s: %s", ticker, e)
        return out

    # ------------------------------------------------------------------
    def fetch_daily(self) -> dict[str, pd.DataFrame]:
        """Daily candles, cached for 5 minutes."""
        now = datetime.now()
        if self._daily_ts and (now - self._daily_ts).total_seconds() < 300:
            return self._daily_cache

        out: dict[str, pd.DataFrame] = {}
        for ticker in TICKERS:
            try:
                df = yf.Ticker(ticker).history(period="1mo", interval="1d")
                if not df.empty:
                    out[ticker] = df
            except Exception as e:
                logger.warning("Daily fetch failed for %s: %s", ticker, e)

        self._daily_cache = out
        self._daily_ts = now
        return out

    # ------------------------------------------------------------------
    def fetch_weekly(self) -> dict[str, pd.DataFrame]:
        """Weekly candles, cached for 30 minutes."""
        now = datetime.now()
        if self._weekly_ts and (now - self._weekly_ts).total_seconds() < 1800:
            return self._weekly_cache

        out: dict[str, pd.DataFrame] = {}
        for ticker in TICKERS:
            try:
                df = yf.Ticker(ticker).history(period="3mo", interval="1wk")
                if not df.empty:
                    out[ticker] = df
            except Exception as e:
                logger.warning("Weekly fetch failed for %s: %s", ticker, e)

        self._weekly_cache = out
        self._weekly_ts = now
        return out

    # ------------------------------------------------------------------
    def fetch_all(self) -> dict:
        return {
            "intraday": self.fetch_intraday(),
            "daily": self.fetch_daily(),
            "weekly": self.fetch_weekly(),
        }
